File: jw-python/src/step0_repl.py
"""Implements step 0 of https://github.com/kanaka/mal - the repl."""


def READ(input_string: str) -> str:
    """Pass though (dummy)."""
    return input_string


def EVAL(ast: str) -> str:
    """Pass though (dummy)."""
    return ast


def PRINT(ast: str) -> None:
    """Print argument (minimal print function)."""
    print(ast)


def rep(input_string: str) -> None:
    """Call read-eval-print on its argument."""
    PRINT(EVAL(READ(input_string)))


def rep_loop() -> None:
    """Repeatedly provide user prompt and passes the input to read-eval-print."""
    while True:
        try:
            rep(input("user> "))
        except EOFError:
            break


if __name__ == "__main__":
    rep_loop()

File: jw-swift/src/step0_repl.swift
// jw-swift implementation of mal
//
// (C) Joe Watson 2020-06-11

func READ(_ input: String) -> String {
    return input
}

func EVAL(_ ast: String) -> String {
    return ast
}

func PRINT(_ ast: String) -> String {
    return ast
}

func rep(_ s: String) -> String {
    return PRINT(EVAL(READ(s)))
}

while true {
    print("user> ", terminator: "")
    if let s = readLine() {
        print(rep(s))
    } else {
        break
    }
}
